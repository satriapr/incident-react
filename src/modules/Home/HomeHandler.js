import { useState, useEffect, useCallback } from 'react';
import { get } from 'lodash';

import IncidentService from '../../services/IncidentService';
import UserService from '../../services/UserService';
import Constant from '../../constants/Constant';

// Home Handler. Contain business logic
const HomeHandler = () => {
  const [incidents, setIncidents] = useState([]);
  const [users, setUsers] = useState([]);

  // Selected user from option dropwdown
  const [selectedUser, setSelectedUser] = useState({});

  // Filter status
  const [selectedStatus, setSelectedStatus] = useState(Constant.ALL);

  // Modal Create
  const [isOpenModal, setIsOpenModal] = useState(false);

  // Trigger getIncidents every filter change
  const getIncidents = useCallback(async () => {
    try {
      const filter = {
        status: selectedStatus === Constant.ALL ? '' : selectedStatus
      };
      const response = await IncidentService.getIncidents(filter);
      setIncidents(get(response, 'data', []));
    } catch (err) {
      throw new Error(get(err, 'message', 'HomeHandler.getIncidents'));
    }
  }, [selectedStatus]);

  useEffect(() => {
    getIncidents();
  }, [getIncidents, selectedStatus]);

  useEffect(() => {
    getUsers();
  }, []);

  const getUsers = async () => {
    try {
      const response = await UserService.getUsers();
      const data = get(response, 'data', []);
      setUsers(data);
      setSelectedUser({
        index: 0,
        data: data[0]
      });
    } catch (err) {
      throw new Error(get(err, 'message', 'HomeHandler.getIncidents'));
    }
  };

  const deleteIncident = async (data) => {
    try {
      await IncidentService.deleteIncident(get(data, '_id', ''));
      getIncidents();
    } catch (err) {
      throw new Error(get(err, 'message', 'HomeHandler.getIncidents'));
    }
  };

  const updateIncidentStatus = async (data, status) => {
    try {
      const newData = { ...data, newStatus: status };
      await IncidentService.updateIncidentStatus(newData);
      getIncidents();
    } catch (err) {
      throw new Error(get(err, 'message', 'HomeHandler.getIncidents'));
    }
  };

  const createIncident = async (data) => {
    try {
      await IncidentService.createIncident(data);
      getIncidents();
    } catch (err) {
      throw new Error(get(err, 'message', 'HomeHandler.getIncidents'));
    }
  };

  const handleCloseModal = (reset) => {
    setIsOpenModal(false);
    reset();
  };

  /* Call API and close modal on submit */
  const handleSubmitModal = (data, reset) => {
    const newData = { ...data, reportedBy: get(selectedUser, 'data._id', '') };
    createIncident(newData);
    setIsOpenModal(false);
    reset();
  };

  /* Select current user from dropdown */
  const handleSelectUser = (selectedIndex) => {
    setSelectedUser({
      index: selectedIndex,
      data: users[selectedIndex]
    });
  };

  const handleSelectStatus = (status) => setSelectedStatus(status);

  return {
    users,
    setUsers,
    getUsers,
    setSelectedUser,
    incidents,
    setIncidents,
    getIncidents,
    deleteIncident,
    createIncident,
    updateIncidentStatus,
    isOpenModal,
    handleCloseModal,
    handleSubmitModal,
    setIsOpenModal,
    handleSelectUser,
    selectedUser,
    setSelectedStatus,
    selectedStatus,
    handleSelectStatus
  };
};

export default HomeHandler;
