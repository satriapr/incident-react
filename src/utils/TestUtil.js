import { shallow } from 'enzyme';
import * as React from 'react';

const singleSnapTest = (wrapper, description) => {
  test(description, () => {
    expect(wrapper).toMatchSnapshot();
  });
};

const testSnapshots = (Component, configs) => {
  describe('snapshots', () => {
    configs.forEach((config) => {
      const { props, description, state } = config;
      const wrapper = shallow(<Component {...props} />);
      if (state) wrapper.setState(state);
      singleSnapTest(wrapper, description);
    });
  });
};

export { testSnapshots };
