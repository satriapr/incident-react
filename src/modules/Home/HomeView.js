import React from 'react';
import {
  createTheme,
  ThemeProvider,
  CssBaseline,
  Box,
  AppBar,
  Toolbar,
  Typography,
  Container,
  Grid,
  Paper,
  Button,
  TextField,
  MenuItem,
  InputLabel
} from '@mui/material';
import { get } from 'lodash';

import HomeHandler from './HomeHandler';
import CreateIncidentModal from './CreateIncidentModal/CreateIncidentModalView';

import CommonTable from '../../components/CommonTable/CommonTableView';
import Constant from '../../constants/Constant';

// Home View. Business logic separated in handler
const Home = () => {
  const {
    users,
    incidents,
    deleteIncident,
    updateIncidentStatus,
    isOpenModal,
    handleCloseModal,
    handleSubmitModal,
    setIsOpenModal,
    handleSelectUser,
    selectedUser,
    selectedStatus,
    handleSelectStatus
  } = HomeHandler();

  const mdTheme = createTheme();

  /* Transform users data to satisfy component props */
  const userOptions = users
    .filter((user) => get(user, 'role'))
    .map((user) => {
      return {
        value: get(user, '_id', ''),
        label: get(user, 'fullName', ''),
        role: get(user, 'role', '')
      };
    });

  const generateUserOptions = () => {
    return userOptions.map((option, index) => {
      return (
        <MenuItem key={get(option, 'value', '')} value={index}>
          {option.label}
        </MenuItem>
      );
    });
  };

  /* ToDo: This status options should be coming from DB */
  const statusOptions = [
    Constant.ALL,
    Constant.OPEN,
    Constant.IN_PROGRESS,
    Constant.RESOLVED
  ];
  const generateStatusOptions = () => {
    return statusOptions.map((option) => {
      return (
        <MenuItem key={option} value={option}>
          {option}
        </MenuItem>
      );
    });
  };

  return (
    <ThemeProvider theme={mdTheme}>
      <Box sx={{ display: 'flex' }}>
        <CssBaseline />
        <AppBar position="absolute">
          <Toolbar>
            <Typography
              component="h1"
              variant="h6"
              color="inherit"
              noWrap
              sx={{ flexGrow: 1 }}
            >
              Incident Dashboard
            </Typography>
            {/* Create button. Only Admin can create */}
            {get(selectedUser, 'data.role') === Constant.ADMIN && (
              <Button
                onClick={() => setIsOpenModal(true)}
                variant="contained"
                color="secondary"
              >
                Create
              </Button>
            )}
          </Toolbar>
        </AppBar>
        <Box
          component="main"
          sx={{
            backgroundColor: (theme) =>
              theme.palette.mode === 'light'
                ? theme.palette.grey[100]
                : theme.palette.grey[900],
            flexGrow: 1,
            height: '100vh',
            overflow: 'auto'
          }}
        >
          <Toolbar />
          <Container maxWidth="lg" sx={{ mt: 4, mb: 4 }}>
            <Grid container spacing={3}>
              <Grid item>
                <InputLabel>Filter by Status</InputLabel>
                <TextField
                  onChange={(e) => handleSelectStatus(e.target.value)}
                  value={selectedStatus || ''}
                  select
                  sx={{ width: 150 }}
                  variant="standard"
                  defaultValue={Constant.ALL}
                >
                  {generateStatusOptions()}
                </TextField>
              </Grid>
              <Grid item>
                {/* User selection dropdown. Just a helper because we don't have auth. */}
                <InputLabel>Selected User</InputLabel>
                <TextField
                  onChange={(e) => handleSelectUser(e.target.value)}
                  value={
                    get(selectedUser, 'index') ||
                    get(selectedUser, 'index') === 0
                      ? get(selectedUser, 'index')
                      : ''
                  }
                  select
                  sx={{ width: 150 }}
                  variant="standard"
                >
                  {generateUserOptions()}
                </TextField>
              </Grid>
              {/* Table List */}
              <Grid item xs={12}>
                <Paper sx={{ p: 2, display: 'flex', flexDirection: 'column' }}>
                  <CommonTable
                    rows={incidents}
                    cells={Constant.CELLS_DATA}
                    handleDeleteCB={deleteIncident}
                    handleAcknowledgeCB={updateIncidentStatus}
                    handleResolveCB={updateIncidentStatus}
                    selectedUser={selectedUser}
                  />
                </Paper>
              </Grid>
            </Grid>
          </Container>
        </Box>
      </Box>

      {/* Create Modal */}
      <CreateIncidentModal
        isOpen={isOpenModal}
        handleCloseCB={(reset) => handleCloseModal(reset)}
        handleSubmitCB={(data, reset) => handleSubmitModal(data, reset)}
        userOptions={userOptions}
      />
    </ThemeProvider>
  );
};

export default Home;
