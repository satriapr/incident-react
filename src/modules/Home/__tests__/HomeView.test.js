import Home from '../HomeView';
import { testSnapshots } from '../../../utils/TestUtil';

describe('Home Component', () => {
  testSnapshots(Home, [
    {
      description: 'Home props',
      props: {}
    }
  ]);
});
