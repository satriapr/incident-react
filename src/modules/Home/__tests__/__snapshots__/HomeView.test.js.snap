// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component snapshots Home props 1`] = `ShallowWrapper {}`;
